### ISSUE TITLE
Macro F1 drops classes with zero true positives; correcting F1 alone can reverse the five-metric ranking

### ISSUE THREAD

**Issue by dnncha** [PYangLab/scMultiBench#3]
I found a second error in `classification_metrics.Rmd`, separate from the specificity error in #2. Classes with no correct predictions are removed from the F1 average rather than counted as zero.

Source checked: commit `0c68f87d554a52e363dbfda2a9d1fa9538eeae9f`, [classification_metrics.Rmd](https://github.com/PYangLab/scMultiBench/blob/0c68f87d554a52e363dbfda2a9d1fa9538eeae9f/evaluation_pipelines/classification/classification_metrics.Rmd). The complete 1,823-byte file was recovered and its Git blob SHA verified as `17c590896591752be5a06098ddf2ac2ac248d419`.

The relevant statements are:

```r
f1_score <- (2 * (precision * sensitivity) / (precision + sensitivity))
f1_score <- mean(f1_score,na.rm = TRUE)
```

When a class has TP=0 but positive ground-truth and predicted support, precision and sensitivity are both zero. The first line produces `0/0 = NaN`; the second drops that class. This is not an ambiguity about classes absent from the test set: every class in the examples below occurs in both the truth and predictions.

For three equally sized classes, predict the first perfectly and swap every observation of the other two classes. The implementation averages `[1, NaN, NaN]` and reports F1=1. The per-class F1 values should be `[1, 0, 0]`, giving macro F1=1/3.

## A complete ranking reversal from changing F1 alone

Here are two constructed confusion matrices on the same 100 observations. Rows are truth, columns are predictions; class supports are 80, 15 and 5 in both cases.

```text
A                  B
74  1  5           79  1  0
 6  0  9            4  7  4
 0  0  5            2  1  2
```

| Metric | A | B |
|---|---:|---:|
| Overall Accuracy | 0.790000 | 0.880000 |
| Average Accuracy / macro recall | 0.641667 | 0.618056 |
| sensitivity | 0.641667 | 0.618056 |
| released `specificity` (still precision) | 0.396053 | 0.680174 |
| released F1 | **0.670833** | 0.634848 |
| corrected F1 | 0.447222 | **0.634848** |

Applying the released five-metric rank-average rule to these two candidates gives **A=1.6, B=1.4** before the F1 correction and **A=1.4, B=1.6** afterwards (larger is better). The separate specificity formula is deliberately unchanged, and duplicate recall is retained. Only the erroneous F1 calculation changes.

These are selected synthetic counterexamples, not the paper's prediction outputs or a corrected publication leaderboard. The ranking fixture was selected by a seeded search, not sampled to estimate the frequency of the error.

## Minimal correction

Replace the per-class F1 expression with the equivalent count formula:

```r
f1_score <- 2 * diag(cm) / (rowSums(cm) + colSums(cm))
```

This is `2*TP/(2*TP+FP+FN)`. For the union of actually observed truth and predicted classes, its denominator is positive, including when TP is zero. The subsequent mean then includes those zero scores. This fixes F1 without changing the specificity calculation tracked in #2.

A compact R check is:

```r
score <- function(cm, corrected = FALSE) {
  recall <- diag(prop.table(cm, 1))
  precision <- diag(prop.table(cm, 2))
  f1 <- 2 * precision * recall / (precision + recall)
  if (corrected) f1 <- 2 * diag(cm) / (rowSums(cm) + colSums(cm))
  c(ACA=mean(recall), OCA=sum(diag(cm))/sum(cm),
    F1=mean(f1, na.rm=TRUE), sensitivity=mean(recall),
    specificity=mean(precision))
}
A <- matrix(c(74,1,5, 6,0,9, 0,0,5), nrow=3, byrow=TRUE)
B <- matrix(c(79,1,0, 4,7,4, 2,1,2), nrow=3, byrow=TRUE)
rank_pair <- function(corrected) {
  values <- rbind(A=score(A, corrected), B=score(B, corrected))
  rowMeans(apply(values, 2, rank, ties.method='max'))
}
stopifnot(isTRUE(all.equal(unname(rank_pair(FALSE)), c(1.6,1.4))))
stopifnot(isTRUE(all.equal(unname(rank_pair(TRUE)), c(1.4,1.6))))
```

## Execution and scope

I executed a Python translation of the R arithmetic, independently checked the corrected values against scikit-learn's label-level macro F1, and verified 499 additional randomly generated confusion matrices, 12 class-permutation checks and 12 count-scaling comparisons. The source checksum was verified. **Native R has not been executed in my current environment**; the R snippet is supplied for that check, not claimed as a completed native run. The Python comparison uses the documented count formula in [scikit-learn's F1 reference](https://scikit-learn.org/stable/modules/generated/sklearn.metrics.f1_score.html).

The F1 part of my suggested replacement in #2 retained this same zero-class problem. That earlier suggestion should not be treated as a complete evaluator fix; the direct count formula above corrects it.

The next publication-level calculation needs the original confusion matrices or prediction labels. This report establishes the implementation error and a possible full ranking reversal, not that a particular published winner changes. I found no earlier report of this exact omission in the repository issues or the targeted web searches performed before submitting.

**Comment by dnncha** [comment]
The native R check is now complete. I ran the original `compute_metrics` function in R 4.4.3 after verifying the full source blob, then changed only the F1 expression. All 206 synthetic fixtures agree with an independent label-level count calculation after correction. The A/B example above reverses from mean ranks 1.6/1.4 to 1.4/1.6 in native R as well.

I also ran the original evaluator directly on the released `data/classification/demo_result/query.csv` and `predict.csv`, both hash-verified at the same commit. That demo has 16,305 observations, 26 classes and no zero-TP classes. Its macro F1 is 0.804534481651323 before and after the correction. It therefore does not trigger this bug; I'm including that result rather than implying every dataset changes.

The successful run, scripts, input manifest and generated results are public:
https://github.com/dnncha/Dnncha/actions/runs/34128607057
https://github.com/dnncha/Dnncha/tree/fde70536c658b11459860e91d96f11fb1da2b051/research/scmultibench-native-r

The earlier native-R limitation is resolved. The separate question of effects on the original publication runs still requires their prediction labels or confusion matrices.
